entity/c/c_api.c

```diff
@@ -0,0 +1,197 @@
+                entity_server_state = HANDSHAKE_2_SENT;
+            }
+        }
+        else if (message_type == SKEY_HANDSHAKE_3)
+        {
+            printf("received session key handshake3!\n");
+            if (entity_server_state != HANDSHAKE_2_SENT)
+            {
+                error_handling("Error during comm init - in wrong state, expected: IDLE, disconnecting...\n");
+            }
+            unsigned int decrypted_length;
+            unsigned char *decrypted = symmetric_decrypt_authenticate(data_buf, data_buf_length, s_key->mac_key, MAC_KEY_SIZE, s_key->cipher_key, CIPHER_KEY_SIZE, AES_CBC_128_IV_SIZE, &decrypted_length);
+            HS_nonce_t hs;
+            parse_handshake(decrypted, &hs);
+            free(decrypted);
+            // compare my_nonce and received_nonce
+            if (strncmp((const char *)hs.reply_nonce, (const char *)server_nonce, HS_NONCE_SIZE) != 0)
+            {
+                error_handling("Comm init failed: server NOT verified, nonce NOT matched, disconnecting...\n");
+            }
+            else
+            {
+                printf("server authenticated/authorized by solving nonce!\n");
+            }
+            printf("switching to IN_COMM\n");
+            entity_server_state = IN_COMM;
+            return s_key;
+        }
+    }
+}
+
+void *receive_thread(void *arguments)
+{
+    while (1)
+    {
+        arg_struct_t *args = (arg_struct_t *)arguments;
+        unsigned char received_buf[1000];
+        unsigned int received_buf_length = read(args->sock, received_buf, sizeof(received_buf));
+        receive_message(received_buf, received_buf_length, args->s_key);
+    }
+}
+
+void receive_message(unsigned char *received_buf, unsigned int received_buf_length, session_key_t *s_key)
+{
+    unsigned char message_type;
+    unsigned int data_buf_length;
+    unsigned char *data_buf = parse_received_message(received_buf, received_buf_length, &message_type, &data_buf_length);
+    if (message_type == SECURE_COMM_MSG)
+    {
+        print_recevied_message(data_buf, data_buf_length, s_key);
+    }
+}
+
+void send_secure_message(char *msg, unsigned int msg_length, session_key_t *s_key, int sock)
+{
+    if (!check_validity(sent_seq_num, s_key->rel_validity, s_key->abs_validity, &st_time))
+    {
+        error_handling("Session key expired!\n");
+    }
+    unsigned char *buf = (unsigned char *)malloc(SEQ_NUM_SIZE + msg_length);
+    memset(buf, 0, SEQ_NUM_SIZE + msg_length);
+    write_in_n_bytes(sent_seq_num, SEQ_NUM_SIZE, buf);
+    memcpy(buf + SEQ_NUM_SIZE, (unsigned char *)msg, msg_length);
+
+    // encrypt
+    unsigned int encrypted_length;
+    unsigned char *encrypted = symmetric_encrypt_authenticate(buf, SEQ_NUM_SIZE + msg_length, s_key->mac_key, MAC_KEY_SIZE, s_key->cipher_key, CIPHER_KEY_SIZE, AES_CBC_128_IV_SIZE, &encrypted_length);
+    free(buf);
+    sent_seq_num++;
+    unsigned char sender_buf[1024]; // TODO: change later.
```

Comment:
What does this TODO mean?

Reply:
**THIS IS RELATED TO THE COMMENT ABOVE OF SENDER_BUF SIZE

Currently the send message does not support dynamic sizes, the max length is shorter than 1024 because this 1024 includes header, encryptions, hmacs etc.

When dynamic buffer sizes are needed, most of the code is implemented in malloc and free, but not when making the sender buf. Do you recommend to change the total connection part to dynamic memory allocation or only when sending messages?

In more details, for example, the auth_hello_reply message's size is dynamic due to the sender_buf(e.g. net1.client) and purpose_buf(e.g. {"group":"Servers"}, {"keyId": 00000000}).
On the other hand, the buffer size when sending handshake1 from client to server is static(Actually changes by encryption scheme, but not implemented yet.).

So, what do you recommend to use when sending bufs? Static of dynamic? If static how should I decide the size of the buffer.

Reply:
My point was to "clarify the TODO". The phrase "change later" adds almost nothing.
What do you need to change? Change the value or logic? Why do you need such a change?

Let's have the static vs. dynamic buffer discussion later. That's not the point of my review comment.

Reply:
Left a detailed comment.